cmd/activator/main.go

```diff
@@ -239,7 +242,35 @@ func main() {
 	}
 	ah = reqLogHandler
 	ah = &activatorhandler.ProbeHandler{NextHandler: ah}
-	ah = &activatorhandler.HealthHandler{HealthCheck: statSink.Status, NextHandler: ah}
+
+	// When we get SIGTERM, start failing readiness probes.
+	servers := map[string]*http.Server{}
+	sigCh := signals.SetupSignalHandler()
+	stopper := sync.Once{}
+	ah = &activatorhandler.HealthHandler{HealthCheck: func() error {
+		select {
+		case <-sigCh:
+			// The first time we get a health check after receiving the SIGTERM,
+			// sleep for a while to let the effects of failing the health check
+			// propagate, drain outstanding requests, and then cancel the context
+			// tearing down the rest of this activator pod.
+			go stopper.Do(func() {
+				// Sleep to let the effects of failing the readiness probe propagate.
+				time.Sleep(30 * time.Second)
+				// Drain outstanding requests, and stop accepting new ones.
+				for _, server := range servers {
+					server.Shutdown(context.Background())
```

Comment:
This will basically never shutdown if you have a streaming connection that never shuts down underneath.
I'd like us to have some upper bound of time we're waiting (since alternative is that after 300s k8s will sigkill the pod).

Reply:
The upper bound is the termination grace period.  I think this is WAI